py: Prevent overflowing size_t when calling mp_seq_multiply.

FelixNumworks opened the pull request:
Previously, it could led to a situation where `m_new(type, BIG_NUMBER)` would call `m_alloc(zero)` because `sizeof(type)*BIG_NUMBER` would return `0`. This change ensures that a MemoryError is raised instead.

EDIT:
The inputs [0]*2\*\*30 or "blue"*2\*\*30 could create a situation where size_t was overflowed and malloc was called with a corrupted value that would then lead to a bad access in memory.

robert-hh commented:
size_t is in most port a 32 bit unsigned quantity. So there are not many ports & mcu's where this will be  a problem. Which port caused the problem for you, which justifies the increase in code size?

FelixNumworks commented:
> size_t is in most port a 32 bit unsigned quantity. So there are not many ports & mcu's where this will be a problem. Which port caused the problem for you, which justifies the increase in code size?

in `objlist.c`, the method `list_binary_op` for the multiplication calls `list_new` which calls `mp_obj_list_init` which calls `m_new` with an arbitrary `n` value.
If the parsed input is `[0]*2**30`, we will have `n = 2**30` so `m_new` will compute `sizeof(mp_obj_t) * 2**30` which yields to `0`. So `m_alloc(0)` will be called, instead of raising a MemoryError, and the list will contain a nullptr that will later on cause a segfault.

robert-hh commented:
Still we are talking about an attempt to allocate memory which usually exceeds the size of the board RAM by far. So it will never succeed, whether or not an error is raised. It only may be harder to trace,

FelixNumworks commented:
> Still we are talking about an attempt to allocate memory which usually exceeds the size of the board RAM by far. So it will never succeed, whether or not an error is raised. It only may be harder to trace,

When `m_malloc(0)` is called, it returns `nullptr`. But `list_binary_op` doesn't check for `nullptr` and then dereferences it. It thus causes a crash of micropython rather than raising a normal python error.

dlech commented:
This is a rather large increase in code size and performance would need checked too since this is adding a multiplication and division operation to every allocation.

Would it be possible to just add an "if alloc size >= heap size then raise memory errror" in one place instead?

robert-hh commented:
> I'm not sure to understand your point I'm sorry.
 
It is good to have  proper error message and not just silent crash. My concern is that this change increases the code size by more that just a little bit, and there are port/board combinations where flash size matters. So getting the same result with a smaller flash increase is better. m_new89 is called at many places in the code. If it's use by list_binary_op is the only place of your concern, it would be better to add a check there.

FelixNumworks committed:
py: Prevent overflowing size_t before callin mp_seq_multiply.

The inputs [0]*2**30 or "blue"*2**30 could create a situation where
size_t was overflowed and malloc was called with a corrupted value that
would then lead to a bad access in memory.

Signed-off-by: Félix Richart <felix.richart@numworks.com>

robert-hh commented:
Do not worry. The Unix coverage check fails almost every time .

dpgeorge commented:
Would it be better (smaller code) to instead put the check at the start of the `mp_seq_multiply` function?  Yes by that point the memory has already been allocated (possibly an incorrect amount), but the `mp_seq_multiply` function is the one writing into that memory so it could do a check to verify that the total size of the destination is within the limits of `UINT_MAX`.

FelixNumworks committed:
py/sequence.c: Factorize check for int overflow.

Signed-off-by: Félix Richart <felix.richart@numworks.com>

FelixNumworks committed:
py/sequence.c: Add an early return.

Signed-off-by: Félix Richart <felix.richart@numworks.com>

FelixNumworks commented:
This is cleaner now I think. Thanks for your help.

dpgeorge commented:
I think there are a lot of other places where a similar overflow can occur, with a multiply overflowing and leading to a memory allocation that's the wrong size (may be 0, may be smaller than requested).

For example:
- in `m_new()`, it multiplies the size of the type by the number of those types to allocate
- in `MP_BINARY_OP_MULTIPLY` operations, working out the number of items

And I think we either fix all of these, or none of them.

If we do fix them, we can do it with a macro that's optional, so it can be disabled to save code size if needed.

Also can use intrinsics in gcc to help with the detection of multiply overflow: https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html

dpgeorge commented:
> At the time I did a first draft of the PR where I was adding a size check inside `m_new`

Do you have your original code for this?

> and they were concerns about increasing the size of the code as well as slowing it too much.

Yes, that's definitely still a concern.  But, looking at this problem again after a year, IMO we need to either fix all cases, or none.  Because there's no reason I can see that makes the sequence multiply more (or less) dangerous/critical than any other version of the bug.

FelixNumworks commented:
@dpgeorge I do still have the old code. It's here: https://github.com/micropython/micropython/pull/15606
